File: backend/internal/lib/either/either.go
package either

import (
	"fmt"
)

func Must[T any](val T, err error) T {
	if err != nil {
		panic(fmt.Errorf("must: %w", err))
	}
	return val
}

func Error[T any](_ T, err error) error {
	if err == nil {
		panic("err is nil")
	}
	return err
}

func Left[T any, U any](val T, _ U) T {
	return val
}

func Right[T any, U any](_ T, val U) U {
	return val
}

func MapLeft[T any, U any, V any](val T, _ U, f func(T) V) V {
	return f(val)
}

func MapRight[T any, U any, V any](_ T, val U, f func(U) V) V {
	return f(val)
}

func FlatMapLeft[T any, U any, V any](val T, _ U, f func(T) (V, error)) (V, error) {
	return f(val)
}

func FlatMapRight[T any, U any, V any](_ T, val U, f func(U) (V, error)) (V, error) {
	return f(val)
}

File: frontend/app/lib/either/either.ts
export type Either<L, R> = Left<L> | Right<R>;

export class Left<L> {
  constructor(public readonly value: L) {}

  isLeft(): this is Left<L> {
    return true;
  }

  isRight(): this is Right<never> {
    return false;
  }

  map<U>(_: (value: never) => U): Either<L, U> {
    return this;
  }

  mapLeft<U>(fn: (value: L) => U): Either<U, never> {
    return new Left(fn(this.value));
  }
}

export class Right<R> {
  constructor(public readonly value: R) {}

  isLeft(): this is Left<never> {
    return false;
  }

  isRight(): this is Right<R> {
    return true;
  }

  map<U>(fn: (value: R) => U): Either<never, U> {
    return new Right(fn(this.value));
  }

  mapLeft<U>(_: (value: never) => U): Either<U, R> {
    return this;
  }
}
